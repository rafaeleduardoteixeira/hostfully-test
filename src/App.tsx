import React from 'react';
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import createdRoutes from './constants/router';
import { Header } from './components/Header/Header';
import { AppContainer } from './App.styles';

function App() {
  return (
    <AppContainer>
      <Header />
      <RouterProvider router={createdRoutes} />
    </AppContainer>
  );
}

export default App;
