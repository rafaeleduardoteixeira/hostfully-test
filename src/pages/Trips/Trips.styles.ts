import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  margin: 20px;
`;

export const TripsTitle = styled.h1`
  font-size: 2rem;
  margin: 20px 0;
`;

export const TripsCardContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
  margin-top: 20px;
`;

export const TripsReturn = styled.img`
  cursor: pointer;
  width: 30px;
  height: 30px;
  z-index: 2;
`;
