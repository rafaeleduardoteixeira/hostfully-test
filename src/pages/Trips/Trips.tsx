import { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { AppContext, AppContextType } from '../../Context/App.context';
import { Container, TripsCardContainer, TripsReturn, TripsTitle } from './Trips.styles';
import { PropertyCard } from '../../components/PropertyCard/PropertyCard';
import moment from 'moment';
import { RentedProperties } from '../../interfaces/rentedProperties';
import { Modal } from '../../components/Modal/Modal';
import { TripDetail } from './components/TripDetail/TripDetail';

export const Trips = () => {
  const { getRentedProperties, handleRemoveRentals, handleEditRentals } = useContext(AppContext) as AppContextType;
  const [clickedTrip, setClickedTrip] = useState<RentedProperties | null>(null);
  const navigate = useNavigate();

  const handleTripPropertyClick = (rentedProperty: RentedProperties) => {
    setClickedTrip(rentedProperty);
  };

  const handleEditTrip = (rentedProperty: RentedProperties) => {
    handleEditRentals(rentedProperty);
    setClickedTrip(null);
  };

  const handleDeleteTrip = (rentedProperty: RentedProperties) => {
    handleRemoveRentals(rentedProperty.id);
    setClickedTrip(null);
  };

  const handleCancel = () => {
    setClickedTrip(null);
  };

  const handleReturn = () => {
    navigate('/');
  };

  return (
    <Container data-testid="Trips">
      <TripsReturn src="/icons/returnArrow.svg" onClick={handleReturn} data-testid="TripReturnHome" />
      <TripsTitle>Trips</TripsTitle>
      <TripsCardContainer>
        {getRentedProperties().map((rentedProperty) => (
          <PropertyCard
            key={rentedProperty.id}
            card={{
              id: rentedProperty.id,
              title: rentedProperty.name,
              subTitle: rentedProperty.propertyName as string,
              description: `${moment(rentedProperty.checkIn, 'America/Los_Angeles').format('ll')} - ${moment(rentedProperty.checkOut, 'America/Los_Angeles').format('ll')}`,
              images: rentedProperty?.propertyImages as string[],
            }}
            onClick={() => handleTripPropertyClick(rentedProperty)}
          />
        ))}
      </TripsCardContainer>
      {clickedTrip && (
        <Modal width="330px">
          <TripDetail
            rentedProperty={clickedTrip}
            handleEditTrip={handleEditTrip}
            handleDeleteTrip={handleDeleteTrip}
            handleCancel={handleCancel}
          />
        </Modal>
      )}
    </Container>
  );
};
