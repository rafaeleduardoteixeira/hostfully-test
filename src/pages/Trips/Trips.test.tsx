import React, { act } from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { Trips } from './Trips';
import { AppContext } from '../../Context/App.context';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe('Trips', () => {
  test('Snapshot test', () => {
    const { asFragment } = render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: () => {
            return [
              {
                id: '1',
                propertyId: '1',
                propertyImages: ['image'],
                propertyName: 'Property 1',
                name: 'Name 1',
                email: 'Email 1',
                checkIn: new Date(),
                checkOut: new Date(),
              },
            ];
          },
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Trips />
      </AppContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Should render Trips', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: () => {
            return [
              {
                id: '1',
                propertyId: '1',
                propertyImages: ['image'],
                propertyName: 'Property 1',
                name: 'Name 1',
                email: 'Email 1',
                checkIn: new Date(),
                checkOut: new Date(),
              },
            ];
          },
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Trips />
      </AppContext.Provider>,
    );
    const TripsElement = screen.getByTestId('Trips');
    expect(TripsElement).toBeInTheDocument();
  });
  test('Should return to home page', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: () => {
            return [
              {
                id: '1',
                propertyId: '1',
                propertyImages: ['image'],
                propertyName: 'Property 1',
                name: 'Name 1',
                email: 'Email 1',
                checkIn: new Date(),
                checkOut: new Date(),
              },
            ];
          },
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Trips />
      </AppContext.Provider>,
    );
    const TripReturnHome = screen.getByTestId('TripReturnHome');
    TripReturnHome.click();
    expect(mockedUsedNavigate).toHaveBeenCalledWith('/');
  });
  test('Should show modal when click on property', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: () => {
            return [
              {
                id: '1',
                propertyId: '1',
                propertyImages: ['image'],
                propertyName: 'Property 1',
                name: 'Name 1',
                email: 'Email 1',
                checkIn: new Date(),
                checkOut: new Date(),
              },
            ];
          },
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Trips />
      </AppContext.Provider>,
    );
    const PropertyCard = screen.getByTestId('PropertyCard');
    act(() => {
      PropertyCard.click();
    });
    const TripDetail = screen.getByTestId('TripDetail');
    expect(TripDetail).toBeInTheDocument();
  });
});
