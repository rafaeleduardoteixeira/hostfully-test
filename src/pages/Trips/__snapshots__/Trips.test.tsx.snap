// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trips Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="sc-braxZu fcOcYJ"
    data-testid="Trips"
  >
    <img
      class="sc-fQpRED gUMSQP"
      data-testid="TripReturnHome"
      src="/icons/returnArrow.svg"
    />
    <h1
      class="sc-gJhJTp khCjDw"
    >
      Trips
    </h1>
    <div
      class="sc-elDIKY jOlUnv"
    >
      <div
        class="sc-dsLQwm eGkpVM"
        data-testid="PropertyCard"
      >
        <img
          alt="Name 1"
          class="sc-iKTcqh kpXgWH"
          src="image"
        />
        <div
          class="sc-gnpbhQ kBQWpX"
        >
          <p
            class="sc-la-DxNn hVkmwp"
          >
            Name 1
          </p>
          <h3
            class="sc-iCZwEW dRsXDN"
          >
            Property 1
          </h3>
          <p
            class="sc-knesRu EoGBD"
          >
            Jul 24, 2024 - Jul 24, 2024
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
